### ISSUE TITLE
Compatibility problems with gdb.Value

### ISSUE THREAD

**Issue by jmpews** [cloudburst/libheap#17]
```
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.04) 7.11.1
```
---
#### Issue: `bin_at`
```
    return int(gdb.parse_and_eval("&((struct malloc_state *) 0x%x).bins[%d]" % \
            (m.address, int((i -1) * 2))).cast(gdb.lookup_type(cast_type)) \
            - offsetof_fd)
```
`m.address` is the type of `gdb.Value` object, so `%x` will crash.

---
#### Issue: `"{:#x}".format(self.val['top'])`
same as above

---

i just replace `m.address` as `str(m.address).split(" ")[0]`.

**Comment by BrieflyX** [comment]
You could just cast `gdb.Value` to `int`, just replace `m.address` to `int(m.address)`.
There are several same problems in the whole script, not working on gdb compiled with python 3.

**Comment by cloudburst** [comment]
I committed a fix for the first one in 5a3b70f7395b359063f44494c319ff4f031e5886

I'm not sure where the second one is occurring.  Do you maybe have an old git checkout?  As I don't see the line "{:#x}".format(self.val['top']) anywhere in current code.  Thanks for the report.

**Comment by jmpews** [comment]
@BrieflyX Yes, i just replace all `gdb.Value`  with `int`. i just found all `gdb.Value` object with `__str__` attribute, so... (最近在研究 heap 相关的, 你平时用 weibo 么? 希望可以多交流.)
@cloudburst https://github.com/cloudburst/libheap/blob/master/prettyprinters.py#L60, most in `prettyprinters.py`.

```
def pretty_print_heap_lookup(val):
    "Look-up and return a pretty printer that can print val."

    val_type = val.type

    # If it points to a reference, get the reference.
    if val_type.code == gdb.TYPE_CODE_REF:
        val_type = val_type.target()

    # Get the unqualified type, stripped of typedefs.
    val_type = val_type.unqualified().strip_typedefs()

    # Get the type name.
    typename = val_type.tag
    if typename == None:
        return None
    elif typename == "malloc_par":
        return malloc_par_printer(val)
    elif typename == "malloc_state":
        return malloc_state_printer(val)
    elif typename == "malloc_chunk":
        return malloc_chunk_printer(val)
    elif typename == "_heap_info":
        return heap_info_printer(val)
    else:
        print(typename)

    # Cannot find a pretty printer for type(val)
    return None

gdb.pretty_printers.append(pretty_print_heap_lookup)
```
`gdb print` is enough for me, so i just `#gdb.pretty_printers.append(pretty_print_heap_lookup)`.

**Comment by BrieflyX** [comment]
@jmpews 不怎么用微博:(，欢迎多多交流~

**Comment by cloudburst** [comment]
I just tried this with gdb 7.11.1 (64bit/python3) on Ubuntu 16.04 LTS and I can't reproduce the pretty printer errors.

(gdb) python print(malloc_state(0xdeadbeef)), etc works fine for me.

Please reopen if you can show a sample command that will trigger the second error.  Thank you.

**Comment by cloudburst** [comment]
@jmpews The pretty printers error is now fixed in 63b3dd2242fcbf42cf59d00fc9e7f65ee9c945b3

Sorry I couldn't originally understand how to reproduce.  I am working on fixing other errors as well.  Thank you for opening an issue!

**Comment by cloudburst** [comment]
@jmpews I've now tested the library rewrite on Ubuntu 14.04 and 16.04 on both 32bit and 64bit.  I think all of the casting errors should be gone now.

Please let me know if you have any more issues with the latest git checkout!
